src/nbl/video/IOpenGL_LogicalDevice.h

```diff
@@ -0,0 +1,692 @@
+    }
+    template <>
+    void IOpenGL_LogicalDeviceBase::SRequestBase<IOpenGL_LogicalDeviceBase::SRequestFlushMappedMemoryRanges>::copyContentsToOwnedMemory(SRequestFlushMappedMemoryRanges& x, void* mem)
+    {
+
+    }
+
+    template <>
+    size_t IOpenGL_LogicalDeviceBase::SRequestBase<IOpenGL_LogicalDeviceBase::SRequestInvalidateMappedMemoryRanges>::neededMemorySize(const SRequestInvalidateMappedMemoryRanges& x)
+    {
+        return x.memoryRanges.size() * sizeof(IDriverMemoryAllocation::MappedMemoryRange);
+    }
+    template <>
+    void IOpenGL_LogicalDeviceBase::SRequestBase<IOpenGL_LogicalDeviceBase::SRequestInvalidateMappedMemoryRanges>::copyContentsToOwnedMemory(SRequestInvalidateMappedMemoryRanges& x, void* mem)
+    {
+
+    }
+*/
+}
+
+// Common base for GL and GLES logical devices
+// All COpenGL* objects (buffers, images, views...) will keep pointer of this type (just to be able to request destruction in destructor though)
+// Implementation of both GL and GLES is the same code (see COpenGL_LogicalDevice) thanks to IOpenGL_FunctionTable abstraction layer
+class IOpenGL_LogicalDevice : public ILogicalDevice, protected impl::IOpenGL_LogicalDeviceBase
+{
+protected:
+    struct SRequest
+    {
+        using params_variant_t = std::variant<
+            SRequestSemaphoreCreate,
+            SRequestEventCreate,
+            SRequestFenceCreate,
+            SRequestFramebufferCreate,
+            SRequestBufferCreate,
+            SRequestBufferViewCreate,
+            SRequestImageCreate,
+            SRequestImageViewCreate,
+            SRequestSamplerCreate,
+            SRequestRenderpassIndependentPipelineCreate,
+            SRequestComputePipelineCreate,
+            SRequestSwapchainCreate,
+
+            SRequest_Destroy<ERT_BUFFER_DESTROY>,
+            SRequest_Destroy<ERT_TEXTURE_DESTROY>,
+            SRequest_Destroy<ERT_FRAMEBUFFER_DESTROY>,
+            SRequest_Destroy<ERT_SWAPCHAIN_DESTROY>,
+            SRequest_Destroy<ERT_SYNC_DESTROY>,
+            SRequest_Destroy<ERT_SAMPLER_DESTROY>,
+            SRequest_Destroy<ERT_PROGRAM_DESTROY>,
+
+            SRequestGetEventStatus,
+            SRequestResetEvent,
+            SRequestSetEvent,
+            SRequestResetFences,
+            SRequestWaitForFences,
+            SRequestFlushMappedMemoryRanges,
+            SRequestInvalidateMappedMemoryRanges
+        >;
+
+        E_REQUEST_TYPE type;
+        params_variant_t params_variant;
+
+        // cast to `RequestParams::retval_t*`
+        void* pretval;
+        // wait on this for result to be ready
+        std::condition_variable cvar;
+        bool ready = false;
+    };
+
+    template <typename FunctionTableType>
+    class CThreadHandler : public system::IThreadHandler<FunctionTableType>
+    {
+        using FeaturesType = typename FunctionTableType::features_t;
+
+        constexpr static inline uint32_t MaxRequestCount = 256u;
```

Comment:
this buffer probably needs to be a lot bigger than 256

Reply:
also why isnt this circular buffer in `IAsync` ?

Reply:
actually why isnt this built off `IAsyncQueue` ?

Reply:
IAsyncQueue is implemented on core:queue. But yeah, ill put this circualr buffer thing to core and get rid of current impl of IAsyncQueue i think

Reply:
> this buffer probably needs to be a lot bigger than 256

maybe, would have to test in practice, best to make the size template param